tests/extmod_hardware/machine_uart_irq_rx.py

```diff
@@ -0,0 +1,78 @@
+# Test machine.UART.IRQ_RX firing for each character received.
+#
+# IMPORTANT: This test requires hardware connections: the UART TX and RX
+# pins must be wired together.
+
+try:
+    from machine import UART
+
+    UART.IRQ_RX
+except (ImportError, AttributeError):
+    print("SKIP")
+    raise SystemExit
+
+import time, sys
+
+byte_by_byte = False
+# Configure pins based on the target.
+if "esp32" in sys.platform:
+    uart_id = 1
+    tx_pin = 4
+    rx_pin = 5
+elif "pyboard" in sys.platform:
+    uart_id = 4
+    tx_pin = None  # PA0
+    rx_pin = None  # PA1
+elif "samd" in sys.platform and "ItsyBitsy M0" in sys.implementation._machine:
+    uart_id = 0
+    tx_pin = "D1"
+    rx_pin = "D0"
+    byte_by_byte = True
+elif "samd" in sys.platform and "ItsyBitsy M4" in sys.implementation._machine:
+    uart_id = 3
+    tx_pin = "D1"
+    rx_pin = "D0"
+elif "samd" in sys.platform and "Seeed" in sys.implementation._machine:
+    uart_id = 0
+    tx_pin = "A4_D4"
+    rx_pin = "A5_D5"
+elif "nrf" in sys.platform:
+    uart_id = 0
+    tx_pin = None
+    rx_pin = None
+elif "renesas-ra" in sys.platform:
+    uart_id = 9
+    tx_pin = None  # P602 @ RA6M2
+    rx_pin = None  # P601 @ RA6M2
+else:
+    print("SKIP")
+    raise SystemExit
+
+
+def irq(u):
+    print("IRQ_RX:", bool(u.irq().flags() & u.IRQ_RX), "data:", u.read(1))
+
+
+text = "1234"
+
+# Test that the IRQ is called for each byte received.
+# Use slow baudrates so that the IRQ has time to run.
+for bits_per_s in (2400, 9600):
+    if tx_pin is None:
+        uart = UART(uart_id, bits_per_s)
+    else:
+        uart = UART(uart_id, bits_per_s, tx=tx_pin, rx=rx_pin)
+
+    uart.irq(irq, uart.IRQ_RX)
+
+    print("write", bits_per_s)
+    if byte_by_byte:
```

Comment:
I suggest to just always write byte-by-byte.  That doesn't really change the test for those ports that are fast enough.

Reply:
But then the test would not reflect the real operation. Better then to skip testing slow devices like SAMD21.

Reply:
Then the .exp file does not match.

Reply:
That's true...

OK, then let's leave it as you have it.